### PR TITLE
bpo-25439: Added additional type checks to urllib.request.Request

### PR DESCRIPTION
**Pull request by CuriousLearner**
Add additional type checks in `urllib.request.Request` to ensure URL is of
type `str`, POST data and it's keys are of type `bytes`.

https://bugs.python.org/issue25439

### CODE REVIEW

File: Doc/library/urllib.request.rst [hunk 1]
```diff
@@ -30,9 +30,9 @@ The :mod:`urllib.request` module defines the following functions:
    Open the URL *url*, which can be either a string or a
    :class:`Request` object.
 
-   *data* must be an object specifying additional data to be sent to the
-   server, or ``None`` if no such data is needed.  See :class:`Request`
-   for details.
+   *data* must be a :term:`bytes-like object <bytes-like object>` specifying
+   additional data to send to the server, or ``None`` if no such data is
+   needed.  See :class:`Request` for details.
```
Comment (vadmium) [inline Doc/library/urllib.request.rst]: Files and other iterables without Content-Length should also be supported since 3.6 (https://bugs.python.org/issue12319). I think it is more helpful to leave this as a summary and include the less-useful details under the Request class.

File: Doc/library/urllib.request.rst [hunk 2]
```diff
@@ -199,7 +199,11 @@ The following classes are provided:
    bytes objects, while ``Transfer-Encoding: chunked`` as specified in
    :rfc:`7230`, Section 3.3.1 will be used to send files and other iterables.
 
-   For an HTTP POST request method, *data* should be a buffer in the
+   For an HTTP POST request method, *data* must be either
+   :term:`bytes-like object <bytes-like object>` or an iterable of it,
+   including a dict with key of this type (this is not supported
+   for HTTPS), and in that case Content-Length value must be specified
```
Comment (vadmium) [inline Doc/library/urllib.request.rst]: Nan’s patch listed supported data types in general, for all HTTP request methods. These types should work for other HTTP methods, not just POST.

I expect the problem with HTTPS is fixed now; see <https://bugs.python.org/issue27340>.

Content-Length is no longer required; read the previous paragraph.

File: Lib/test/test_urllib2.py [hunk 3]
```diff
@@ -1922,7 +1918,7 @@ def test_url_fullurl_get_full_url(self):
                 'http://docs.python.org/library/urllib2.html#OK',
                 'http://www.python.org/?qs=query#fragment=true']
         for url in urls:
-            req = Request(url)
+            req = Request(url, None)
```
Comment (vadmium) [inline Lib/test/test_urllib2.py]: Why was this added? None should be the default, if no _data_ argument is given.

**Review by vadmium — COMMENTED** [review]
FWIW some of my earlier comments also still apply, in particular dict-of-bytearray <https://bugs.python.org/issue25439#msg253815> and not mentioning the _dict_ special case <https://bugs.python.org/review/25439/diff/15859/Doc/library/urllib.request.rst#newcode181>

**Comment by CuriousLearner** [comment]
@vadmium Can you have a look now?

**Review by vadmium — COMMENTED** [review]
I had a look, but I don’t like the documentation changes, and I don’t understand why you want to specifically whitelist _bytes_ and _memoryview_ dictionary keys, yet drop support for other kinds of keys.

**Comment by CuriousLearner** [comment]
Hey @vadmium !

I whitelisted `bytes` and `memoryview` solely based on examples. Do you think we can improve this somehow? What do you think we can improve in the documentation?

**Comment by CuriousLearner** [comment]
Hello @vadmium Do you have any suggestions on this?

**Review by MaxwellDupre — APPROVED** [review]
Run tests passed for both urllib and 2.
Looks ok.

**Review by iritkatriel — DISMISSED** [review]
This has a merge conflict now.

**Comment by CuriousLearner** [comment]
Hi @iritkatriel 

Seems like Ezio updated the branch with latest changes in master. Please have a look and let me know if there is anything I can do to get this merged :) 

Thanks!
